genum-derive/src/lib.rs

```diff
@@ -0,0 +1,295 @@
+                ItemAttribute::Name(n) => value_name = n,
+                ItemAttribute::Nick(n) => value_nick = n,
+            }
+        );
+
+        let value_name = format!("{}\0", value_name);
+        let value_nick = format!("{}\0", value_nick);
+
+        n += 1;
+        quote_spanned! {v.span()=>
+            gobject_sys::GEnumValue {
+                value: #enum_name::#name as i32,
+                value_name: #value_name as *const _ as *const _,
+                value_nick: #value_nick as *const _ as *const _,
+            },
+        }
+    });
+    (
+        quote! {
+            #(#recurse)*
+        },
+        n,
+    )
+}
+
+// Parse enum attribute such as:
+// #[genum(type_name = "TestAnimalType")]
+fn parse_type_name(input: &DeriveInput) -> Result<String> {
+    let meta = match find_genum_meta(&input.attrs)? {
+        Some(meta) => meta,
+        _ => bail!("Missing 'genum' attribute"),
+    };
+
+    let meta = match meta.nested.first() {
+        Some(meta) => meta,
+        _ => bail!("Missing meta 'type_name'"),
+    };
+
+    match parse_enum_attribute(&meta)? {
+        EnumAttribute::TypeName(n) => Ok(n),
+    }
+}
+
+fn impl_genum(input: &syn::DeriveInput) -> TokenStream {
+    let name = &input.ident;
+
+    let enum_variants = match input.data {
+        Data::Enum(ref e) => &e.variants,
+        _ => abort_call_site!("GEnum only supports enums"),
+    };
+
+    let gtype_name = match parse_type_name(&input) {
+        Ok(v) => v,
+        Err(e) => abort_call_site!(
+            "{}: derive(GEnum) requires #[genum(type_name = \"EnumTypeName\")]",
+            e
+        ),
+    };
+    let get_type = format_ident!("{}_get_type", name.to_string().to_snake_case());
+    let from_glib = gen_from_glib(name, enum_variants);
+    let (genum_values, nb_genum_values) = gen_genum_values(name, enum_variants);
+
+    quote! {
+        impl glib::translate::ToGlib for #name {
+            type GlibType = i32;
+
+            fn to_glib(&self) -> i32 {
+                *self as i32
+            }
+        }
+
+        impl glib::translate::FromGlib<i32> for #name {
+            fn from_glib(value: i32) -> Self {
+                #from_glib
+                unreachable!();
+            }
+        }
+
+        impl<'a> glib::value::FromValueOptional<'a> for #name {
+            unsafe fn from_value_optional(value: &glib::Value) -> Option<Self> {
+                Some(glib::value::FromValue::from_value(value))
+            }
+        }
+
+        impl<'a> glib::value::FromValue<'a> for #name {
+            unsafe fn from_value(value: &glib::Value) -> Self {
+                use glib::translate::ToGlibPtr;
+
+                glib::translate::from_glib(gobject_sys::g_value_get_enum(value.to_glib_none().0))
+            }
+        }
+
+        impl glib::value::SetValue for #name {
+            unsafe fn set_value(value: &mut glib::Value, this: &Self) {
+                use glib::translate::{ToGlib, ToGlibPtrMut};
```

Comment:
`use` in macro always makes me a bit nervous. It could cause errors if the same type is already imported outside the macro, for example. Or in case of traits if the used type also implements a function with the given name then not the one of the trait would be used.

Generally it's nicer to directly call the trait methods `glib::translate::ToGlib::to_glib(this)`, `glib::translate::ToGlibPtrMut(value)`